Use full chain of member expressions when generating display names.

mihaip opened the pull request:
Assumed to be namespaced names (with the exception of "exports", which is special-cased)

syranide commented:
The current behavior is how JavaScript generally implies display names though IIRC? You're always free to provide your own displayName if you feel the one generated by JSX is insufficient.

mihaip commented:
@syranide I don't think JS ever generates display names for property assignments, `foo.bar = function() {...};` will end up with an anonymous function. Also, my understanding is that `displayName` is meant for debugging purposes only, so it does not need to follow JS semantics.

spicyj commented:
What do you think about doing something like this?

```js
/**
 * If `expr` is an Identifier or MemberExpression node made of identifiers and
 * dot accesses, return a list of the identifier parts. Other nodes return null.
 *
 * Examples:
 *
 * MyComponent -> ['MyComponent']
 * namespace.MyComponent -> ['namespace', 'MyComponent']
 * namespace['foo'] -> null
 * namespace['foo'].bar -> ['bar']
 */
function flattenIdentifierOrMemberExpression(expr) {
  if (expr.type === Syntax.Identifier) {
    return [expr.name];
  } else if (expr.type === Syntax.MemberExpression) {
    if (!expr.computed && expr.property.type === Syntax.Identifier) {
      var flattenedObject = flattenIdentifierOrMemberExpression(expr.object);
      if (flattenedObject) {
        flattenedObject.push(expr.property.name);
        return flattenedObject;
      } else {
        return [expr.property.name];
      }
    }
  }
  return null;
}

...

var displayNamePath = flattenIdentifierOrMemberExpression(left);
if (displayNamePath) {
  if (displayNamePath.length > 1 && displayNamePath[0] === 'exports') {
    displayNamePath.shift();
  }
  addDisplayName(displayNamePath.join('.'), right, state);
}
```

Not tested, but it feels little cleaner to me.

Also, can you add a test for the `exports` behavior too?

spicyj commented on vendor/fbtransform/transforms/reactDisplayName.js:
```diff
@@ -72,11 +73,40 @@ function visitReactDisplayName(traverse, object, path, state) {
+      var memberExpression = left;
+      var displayNamePath = [];
+      while (memberExpression) {
+        if (memberExpression.property.type === Syntax.Identifier) {
```
(You need to check `memberExpression.computed` here or else you'll catch `foo[bar]`.)

mihaip commented:
Updated with your suggestion (worked out of the box).

There was already a test for the `exports` behavior, that did not need to be added.

I also changed the ordering a bit, we were previously computing the candidate display name before we were actually checking that this was a `React.createClass` call. Now that display name computation is more expensive, I've re-ordered such that we only do it if we actually need to.

spicyj commented:
Thank you!
